### PR TITLE
Allow an organisation admin to upgrade a group

### PR DESCRIPTION
**Pull request by stephencdaly**
### What problem does this pull request solve?

Trello card: https://trello.com/c/XCLShuwW/1483-upgrade-groups-org-admin-can-upgrade-a-group-without-a-request

Adds a link to the the notification banner for a trial group to upgrade a group when the logged in user is an organisation admin or super admin.

When the link is clicked, `/groups/:external_id/upgrade` is visited which renders a page to confirm upgrading the group. If "Yes" is selected, the group is upgraded and an email is sent to all group admins for the group.

Screenshot 2024-04-26 at 09 29 53

Screenshot 2024-04-26 at 09 29 57

Screenshot 2024-04-26 at 09 30 01

Screenshot 2024-04-26 at 09 30 06

Screenshot 2024-04-26 at 09 45 38

### Things to consider when reviewing

- Ensure that you consider the wider context.
- Does it work when run on your machine?
- Is it clear what the code is doing?
- Do the commit messages explain why the changes were made?
- Are there all the unit tests needed?
- Has all relevant documentation been updated?

### CODE REVIEW

File: app/controllers/groups_controller.rb [hunk 1]
```diff
@@ -56,6 +56,18 @@ def confirm_upgrade
     @confirm_upgrade_form = Groups::ConfirmUpgradeForm.new
   end
 
+  def upgrade
+    authorize @group
+
+    @confirm_upgrade_form = Groups::ConfirmUpgradeForm.new(confirm_upgrade_form_params)
+    return render :confirm_upgrade unless @confirm_upgrade_form.valid?
```
[the remark this answers is not in the document]
Comment (stephencdaly) [reply app/controllers/groups_controller.rb]: done

File: spec/views/groups/show.html.erb_spec.rb [hunk 2]
```diff
@@ -4,11 +4,17 @@
   let(:current_user) { create :user }
   let(:forms) { [] }
   let(:group) { create :group, name: "My Group" }
+  let(:upgrade) { false }
```
[the remark this answers is not in the document]
Comment (stephencdaly) [reply spec/views/groups/show.html.erb_spec.rb]: done

File: config/locales/en.yml [hunk 3]
```diff
@@ -339,6 +339,10 @@ en:
     status_caption:
       active: Active group
       trial: Trial group
+    success_messages:
+      create: Group was successfully created.
+      update: Group was successfully updated.
```
[the remark this answers is not in the document]
Comment (stephencdaly) [reply config/locales/en.yml:345]: As discussed this was questioning whether these success messages should be displayed at all when a group is created/updated. Agreed to ask this question separate to this team, and can remove in a different PR if they shouldn't exist.

File: spec/form_objects/groups/confirm_upgrade_form_spec.rb [hunk 4]
```diff
@@ -1,15 +1,60 @@
...
+          create(:membership, user: editor_user, group:, role: :editor)
+        end
+      end
+      let(:delivery) { double }
+
+      before do
+        allow(GroupUpgradedMailer).to receive(:group_upgraded_email)
+                                        .with(upgraded_by_user: current_user, to_email: anything, group:, group_url: group_url(group, host:))
+                                        .and_return(delivery)
+        allow(delivery).to receive(:deliver_now).with(no_args)
```
Comment (SamJamCul) [inline spec/form_objects/groups/confirm_upgrade_form_spec.rb:45]: Haven't seen the `.with(no_args)` before but I like it

File: app/form_objects/groups/confirm_upgrade_form.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
+module Groups
+  class ConfirmUpgradeForm < ConfirmActionForm
+    include Rails.application.routes.url_helpers
+
+    attr_accessor :group, :current_user, :host
+
+    def submit
+      group.active!
+      send_notification_emails
+    end
```
Comment (SamJamCul) [inline app/form_objects/groups/confirm_upgrade_form.rb:10]: Did you want to pass in the submit-specific values in this PR, or in a later one that covers all our form objects?

**Review by SamJamCul — APPROVED** [review]
Looking good!

**Review by SamJamCul — APPROVED** [review]

**Pull request merged**